### PR TITLE
LV-RUBY 51

### PR DESCRIPTION
**Pull request by oleksandrklymko**
## JIRA ticket

[LVRUBY-51](http://ssu-jira.softserveinc.com/secure/RapidBoard.jspa?rapidView=979&view=planning&selectedIssue=LVRUBY-51)

## Code reviewers

- [x] @AbdulRustam 
- [ ] @omaik 

## Summary of issue

As a user, I want to add avatar in my profile, so everybody can see my photo.
As a user, I want to update avatar in my profile.

## Summary of change

I added a possibility to add avatar to profile and possibility to update avatar.

## Review checklist

- [ ] Rubocop doesn't raise any warnings and offenses
- [ ] New features are covered with new specs
- [ ] All specs pass

### CODE REVIEW

**Comment by vikabelous** [comment]
**Code Coverage:**
:green_heart: **97.16%** covered at **3.17** hits/line

File: app/controllers/users_controller.rb [hunk 1]
```diff
@@ -0,0 +1,23 @@
...
+  def edit; end
+
+  def show; end
+
+  def update
+    if current_user.update(user_params)
+      flash.now[:notice] = 'You have updated avatar'
+      render 'show'
+    else
+      flash.now[:alert] = 'Wrong format'
```
Comment (omaik) [inline app/controllers/users_controller.rb]: better use what `current_user.errors` says

File: Gemfile [hunk 2]
```diff
@@ -6,7 +6,8 @@ git_source(:github) { |repo| "https://github.com/#{repo}.git" }
 ruby '2.6.2'
 
 gem 'bootsnap', '>= 1.1.0', require: false
-gem 'bootstrap', '~> 4.3.1'
+gem 'bootstrap', '4.1.1'
```
Comment (omaik) [inline Gemfile]: I guess you did it by mistake

File: app/views/users/_form.html.haml [hunk 3]
```diff
@@ -0,0 +1,8 @@
+= form_for current_user, html: { multipart: true } do |f|
```
Comment (omaik) [inline app/views/users/_form.html.haml]: why do you need a form with no input elements in?

File: spec/controllers/users_controller_spec.rb [hunk 4]
```diff
@@ -0,0 +1,60 @@
...
+  let(:wrong_params) { { avatar: Rack::Test::UploadedFile.new('spec/test_zip.zip', 'image/jpeg') } }
+
+  before { sign_in user_without_avatar }
+
+  describe '#update' do
+    it 'is updates avatar for user without avatar' do
+      avatar_before = user_without_avatar.avatar
+      put :update, params: { id: user_without_avatar.id, user: correct_params }
+      user_without_avatar.reload
+      expect(user_without_avatar.avatar).not_to eq(avatar_before)
```
Comment (omaik) [inline spec/controllers/users_controller_spec.rb]: simply just check `current_user.avatar?` as you do in Helper.

File: spec/controllers/users_controller_spec.rb [hunk 5]
```diff
@@ -0,0 +1,60 @@
...
+    end
+
+    it 'is updates avatar for user with avatar' do
+      avatar_before = user_with_avatar.avatar
```
Comment (omaik) [inline spec/controllers/users_controller_spec.rb]: you could something like
```
expect do
  put :update, params: { id: user_with_avatar.id, user: correct_params }
end.to change { user_with_avatar.reload.avatar }
```

File: spec/controllers/users_controller_spec.rb [hunk 6]
```diff
@@ -0,0 +1,60 @@
...
+    ))
+  end
+
+  let(:correct_params) { { avatar: Rack::Test::UploadedFile.new('spec/test1.jpg', 'image/jpeg') } }
+
+  let(:wrong_params) { { avatar: Rack::Test::UploadedFile.new('spec/test_zip.zip', 'image/jpeg') } }
+
+  before { sign_in user_without_avatar }
+
+  describe '#update' do
```
Comment (omaik) [inline spec/controllers/users_controller_spec.rb]: please group the tests using context it is hard to navigate through them 
smt like
```
context 'wrong input' do
  # tests with wrong input
end
```
etc

File: spec/helpers/users_helper_spec.rb [hunk 7]
```diff
@@ -0,0 +1,30 @@
...
+
+  let(:user_with_avatar) do
+    create(:user, avatar: Rack::Test::UploadedFile.new(
+      'spec/test3.jpeg', 'image/jpeg'
+    ))
+  end
+
+  before { sign_in user }
+
+  describe '#user_avatar_url default url' do
```
Comment (omaik) [inline spec/helpers/users_helper_spec.rb]: `describe` is used to indicate the method you are testing.  for different cases use `context`

**Comment by vikabelous** [comment]
**Code Coverage:**
:green_heart: **97.16%** covered at **3.19** hits/line

**Comment by vikabelous** [comment]
**Code Coverage:**
:green_heart: **97.16%** covered at **3.19** hits/line

**Comment by vikabelous** [comment]
**Code Coverage:**
:green_heart: **97.34%** covered at **5.01** hits/line

File: app/controllers/users_controller.rb [hunk 8]
```diff
@@ -0,0 +1,23 @@
+# frozen_string_literal: true
+
+class UsersController < ApplicationController
+  def edit; end
+
+  def show; end
+
+  def update
+    if current_user.update(user_params)
+      flash.now[:notice] = 'You have updated avatar'
```
Comment (omaik) [inline app/controllers/users_controller.rb]: it is general update, please use more abstract message, and add it to I18n

File: app/controllers/users_controller.rb [hunk 9]
```diff
@@ -0,0 +1,23 @@
...
+      render 'show'
+    else
+      flash.now[:alert] = current_user.errors.messages[:avatar].first
```
Comment (omaik) [inline app/controllers/users_controller.rb]: please show all messages, not only avatar

File: app/views/users/_form.html.haml [hunk 10]
```diff
@@ -0,0 +1,7 @@
+- if current_user.errors.any?
```
Comment (omaik) [inline app/views/users/_form.html.haml]: bad name for partial, it is not a form

File: spec/controllers/users_controller_spec.rb [hunk 11]
```diff
@@ -0,0 +1,63 @@
...
+        end.to change { user_without_avatar.reload.avatar }
+      end
+
+      it 'is updates avatar for user with avatar' do
+        expect do
+          put :update, params: { id: user_with_avatar.id, user: correct_params }
+        end.to change { user_with_avatar.reload.avatar }
+      end
+
+      it 'is updates avatar and rendernig show page' do
```
Comment (omaik) [inline spec/controllers/users_controller_spec.rb]: write in the message what you test `renders show page` would be enough. Same for test below

File: app/views/users/show.html.haml [hunk 12]
```diff
@@ -0,0 +1,14 @@
...
+  .row
+    .col-12
+      .card
+        .card-body
+          .card-title.mb-4
+            .d-flex.justify-content-start
+              .image-container
+                = image_tag user_avatar_url, alt: 'Avatar' , class: 'user-img'
+                %br
+                  = link_to "Update", edit_user_path(current_user.id), method: :get , class: 'btn'
```
Comment (omaik) [inline app/views/users/show.html.haml]: use i18n for all hardcoded strings

**Comment by vikabelous** [comment]
**Code Coverage:**
:green_heart: **97.34%** covered at **5.01** hits/line

**Comment by vikabelous** [comment]
**Code Coverage:**
:green_heart: **97.34%** covered at **5.01** hits/line

**Comment by vikabelous** [comment]
**Code Coverage:**
:green_heart: **97.34%** covered at **5.01** hits/line

**Pull request merged**